Initial implementation (GSI-1817)

TheByronHimes opened the pull request.

Cito commented on src/uos/adapters/inbound/__init__.py:
```diff
@@ -0,0 +1,17 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Subpackage defining a inbound/primary/driving adapters according to the
```
```suggestion
"""Subpackage defining inbound/primary/driving adapters according to the
```

Cito commented on src/uos/adapters/inbound/event_sub.py:
```diff
@@ -0,0 +1,69 @@
+        await self._upload_orchestrator.upsert_file_upload_box(file_upload_box=update)
+
+    async def deleted(self, resource_id: str) -> None:
+        """Consume a deleted FileUploadBox event and remove from own storage"""
```
The docstring should be adapted if we ignore it.

Maybe we should make the file upload ID nullable and set it to null in this case, to indicate that the box was deleted. We can still keep the other fields (size and count) at their latest value.

mephenor replied:
Without an ID, is the remaining information even accessible?

Cito replied:
No, but we can still keep its latest state as info.

I would either handle it by e.g. removing the file upload box id (which is completely useless if the box is deleted) or throw an error (if it should never happen), not just ignore it.

TheByronHimes replied:
We don't currently have a mechanism that would delete a FUB through the outbox channel. I mean that the only way to delete FUBs right now is to manually delete them through mongo express or something. So no FUB deletion events are ever going to come in until that is implemented. Likewise, there is no deletion path for RDUBs.
We probably should add this pathway in the near future, but it seriously is not crucial right now and doing so would definitely mean missing this release window.
So for now, I think ignoring it with a log statement is more appropriate than an error because what are we going to do with that error anyway? There is no way for that event to get generated.

TheByronHimes replied:
But I will fix the doc string

Cito replied:
Ok. Maybe we shouldn't overthink this and revisit if we feel the need to clean up the old boxes.

Cito commented on src/uos/adapters/inbound/fastapi_/configure.py:
```diff
@@ -0,0 +1,58 @@
+        version=__version__,
+        description="A service providing a web-accessible management and auth layer"
+        + " over upload-path file services.",
+        tags=[{"name": "UploadOrchestrationService"}],
```
```suggestion
        tags=[{"name": "ResearchDataUploadBoxes"}],
```

mephenor replied:
Other services also have their own name here

Cito replied:
Which does not make much sense, since the name of the service is already in the OpenAPI title.

For grouping it makes much more sense to refer to the things handled by the endpoint. In this service, providing two tags like "ResearchDataUploadBox" and "Grant" (or maybe just "boxes" and "grants") would make more sense. The tags need to be also specified properly at the corresponding endpoints of course.

Btw, it is also possible to add a description, not just a name. And you can have multiple tags. For instance, the auth service has tags for grants and claims, and grants can have additional tags upload and download.

TheByronHimes replied:
It would be good if there was some agreement on exactly what should and should not be included in these tags going forward :)

Cito replied:
For now, I think it's sufficient to just use some short names (not descriptions) for the categories of things that the service handles, so that we have nice grouping in the interactive API docs.

If we ever create an overarching API documentation, we can start thinking more about consistent naming conventions and adding descriptions.

Cito commented on src/uos/adapters/inbound/fastapi_/dummies.py:
```diff
@@ -0,0 +1,44 @@
+]
+
+auth_provider = DependencyDummy("auth_provider")
+upload_orchestrator_port = DependencyDummy("work_package_repo_port")
```
```suggestion
upload_orchestrator_port = DependencyDummy("upload_orchestrator_port")
```

Cito commented on src/uos/adapters/inbound/fastapi_/routes.py:
```diff
@@ -0,0 +1,373 @@
+
+
+@router.get(
+    "/boxes",
```
Good that you added this endpoint already. It was not mentioned in the PR description and epic spec.

TheByronHimes replied:
Actually, I think it was mentioned in the epic spec but it wasn't one of the main endpoints so I only added it at the end.

Cito commented on src/uos/adapters/inbound/fastapi_/routes.py:
```diff
@@ -0,0 +1,373 @@
+@TRACER.start_as_current_span("routes.get_research_data_upload_boxes")
+async def get_research_data_upload_boxes(
+    upload_service: UploadOrchestratorDummy,
+    auth_context: UserAuthContext,
```
Would be good to have filters (query params) for `title` and `locked` (false/true, none=both), and ideally also `changed_before` and `changed_after`.

The results should sorted by `locked` (first unlocked, then locked), then by `last_changed` descending, then by `id`. This will give a stable order by relevance.

Cito replied:
Also, if the user is not a data steward, the results should be filtered to only show boxes that the user has an upload grant for. This should also be mentioned in the docstring.

Cito commented on src/uos/adapters/inbound/fastapi_/routes.py:
```diff
@@ -0,0 +1,373 @@
+    sort: Annotated[
+        SortOrder,
+        Query(description="Sort order(s) that shall be used when sorting results"),
+    ] = SortOrder.ASCENDING,
```
I don't think we need that if we use the order by relevance.

Either remove it or make it more functional, e.g. allow allow `sort` to be an optional field name (optional prefixed with +/- to indicate direction), e.g. `-title` to order by title descending. We would still always use `locked`, `-last_changed`, `id` as additional sort params.

TheByronHimes replied:
I will remove it. BTW, "relevance" is irrelevant here since there's no text search (I think that's overkill for this round). As pagination crops up more, it feels like it would be nice for a drop in solution. maybe something to revisit later

Cito replied:
With order by relevance I meant the default order by locked and then last change, not some kind of relevance for keyword based searches. Because as a data steward, the most relevant boxes are probably the open ones that have last been touched, not some old, closed boxed from years ago.

Cito commented on src/uos/adapters/inbound/fastapi_/routes.py:
```diff
@@ -0,0 +1,373 @@
+    upload_service: UploadOrchestratorDummy,
+    auth_context: UserAuthContext,
+):
+    """Get details of a specific upload box."""
```
Docstring should mention that users get an access error if they do not have an access grant for the box.

Or from a security point of view we could also return a BoxNotFoundError in this case. This would match the behavior of the list get, where these boxes are simply not returned either, as if they would not exist.

TheByronHimes replied:
I went with the latter

Cito commented on src/uos/adapters/inbound/fastapi_/routes.py:
```diff
@@ -0,0 +1,373 @@
+):
+    """Grant upload access to a user. Requires Data Steward role."""
+    # Check if user has Data Steward role
+    if not check_data_steward_role(auth_context):
```
Might make sense to use a stricter auth context for such end points, like [here](https://github.com/ghga-de/auth-service/blob/main/src/auth_service/rest/auth.py).

TheByronHimes replied:
Good idea, wieder was gelernt

Cito commented on src/uos/adapters/inbound/fastapi_/routes.py:
```diff
@@ -0,0 +1,373 @@
+    auth_context: UserAuthContext,
+) -> None:
+    """Revoke an upload access grant."""
+    if not check_data_steward_role(auth_context):
```
see above

Cito commented on src/uos/adapters/inbound/fastapi_/routes.py:
```diff
@@ -0,0 +1,373 @@
+    """Get upload access grants.
+
+    You can filter the grants by user ID, IVA ID, and box ID
+    and by whether the grant is currently valid or not.
```
Again, might make sense to get a list sorted by relevance (valid desc, box id, user id, iva id, grant id).

Cito commented on src/uos/adapters/outbound/audit.py:
```diff
@@ -0,0 +1,102 @@
+        )
+
+    # TODO: How should we log reads? If we log every result from search, we could
+    #  end up with a ton of meaningless logs
```
I think they shouldn't be logged here.

Cito commented on src/uos/adapters/outbound/event_pub.py:
```diff
@@ -0,0 +1,65 @@
+class EventPubTranslator(EventPublisherPort):
+    """A translator according to  the triple hexagonal architecture implementing
+    the EventPublisherPort.
+    """
```
```suggestion
    """A hexkit translator for publishing to the audit log."""
```

Cito commented on src/uos/adapters/outbound/http.py:
```diff
@@ -0,0 +1,373 @@
+    """Config parameters for managing upload access grants."""
+
+    access_url: str = Field(
+        ...,
```
Make this a HttpURL.

Cito commented on src/uos/adapters/outbound/http.py:
```diff
@@ -0,0 +1,373 @@
+    """Config parameters for interacting with the UCS."""
+
+    # maybe this should be a WKVS call? solve later
+    ucs_url: str = Field(
```
see above

Cito commented on src/uos/adapters/outbound/http.py:
```diff
@@ -0,0 +1,373 @@
+class UCSApiConfig(BaseSettings):
+    """Config parameters for interacting with the UCS."""
+
+    # maybe this should be a WKVS call? solve later
```
You mean getting all the API URLs from WKVS? Might be a good idea. But WKVS could become a single point of failure and we would need to ensure to call it first, or call it lazily, which is more overhead. Maybe we should add something to the service commons to enable this? Can you add a ticket to explore that and remove the comment here?

TheByronHimes replied:
I think you make a good point. Configuration is relatively easy to maintain and can be updated quickly. Let's stick with that. It makes sense for the connector, because there configuration is maintained by the end user

Cito commented on src/uos/adapters/outbound/http.py:
```diff
@@ -0,0 +1,373 @@
+                    "response_text": response.text,
+                },
+            )
+            raise self.AccessAPIError("Failed to grant upload access.")
```
```suggestion
        log.error(
            "Failed to revoke upload access for grant ID %s.",
            grant_id,
            extra={
                "grant_id": grant_id,
                "status_code": response.status_code,
                "response_text": response.text,
            },
        )
        raise self.AccessAPIError("Failed to grant upload access.")
```

Cito commented on src/uos/adapters/outbound/http.py:
```diff
@@ -0,0 +1,373 @@
+                        "response_body": response.text,
+                    },
+                )
+                raise self.AccessAPIError("Failed to check box access.")
```
```suggestion
            if response.status_code in (403, 404):
                return False
            log.error(
                "Unexpected response when checking box access for user %s and box %s.",
                user_id,
                box_id,
                extra={
                    "user_id": user_id,
                    "box_id": box_id,
                    "status_code": response.status_code,
                    "response_body": response.text,
                },
            )
            raise self.AccessAPIError("Failed to check box access.")
```

Cito commented on src/uos/core/crypt.py:
```diff
@@ -0,0 +1,42 @@
+_re_pem_public = re.compile("-----(BEGIN|END) CRYPT4GH PUBLIC KEY-----")
+
+
+def validate_public_key(key: str) -> str:
```
do we need this?

TheByronHimes replied:
nope, copied it from the WPS because I wasn't sure how it all worked. forgot to take a second look after all the edits

Cito commented on src/uos/core/models.py:
```diff
@@ -0,0 +1,258 @@
+
+
+class FileUploadBox(BaseModel):
+    """A class representing a box that bundles files belonging to the same upload."""
```
docstrings should mention that these are maintained externally by the UCS

Cito commented on src/uos/core/models.py:
```diff
@@ -0,0 +1,258 @@
+
+
+class FileUpload(BaseModel):
+    """A File Upload."""
```
see above

Cito commented on src/uos/core/models.py:
```diff
@@ -0,0 +1,258 @@
+    changed_by: UUID4 = Field(
+        ..., description="ID of the user who performed the latest change"
+    )
+    file_upload_box_id: UUID4 = Field(..., description="The ID of the file upload box.")
```
```suggestion
    file_upload_box_id: UUID4 = Field(..., description="The ID of the corresponding file upload box.")
```

Cito commented on src/uos/core/models.py:
```diff
@@ -0,0 +1,258 @@
+
+
+class ClaimValidity(BaseModel):
+    """Start and end dates for validating claims."""
```
```suggestion
class GrantValidity(BaseModel):
    """Start and end dates for validating grants."""
```

Cito replied:
We're working on the level of "grants". The claims are a more low level and used only internally (implementation detail).

TheByronHimes replied:
Thanks for the clarification.

Cito commented on src/uos/core/models.py:
```diff
@@ -0,0 +1,258 @@
+class GrantAccessRequest(BaseModel):
+    """Request model for granting upload access to a user."""
+
+    validity: ClaimValidity
```
```suggestion
    validity: GrantValidity
```

Cito replied:
Maybe flatten this, so it looks it's more similar to the UploadGrant.

TheByronHimes replied:
Done!

Cito commented on src/uos/adapters/outbound/http.py:
```diff
@@ -0,0 +1,373 @@
+            raise self.AccessAPIError("Failed to check box access.") from err
+
+
+class UCSClient(UCSClientPort):
```
Maybe call this FileBox client? The fact that this is maintained by the UCS is an implementation detail that should be abstracted away by this adapter.

Cito commented on src/uos/core/orchestrator.py:
```diff
@@ -0,0 +1,392 @@
+        3. Emits events and audit records
+
+        Returns:
+            The UUID of the newly created upload box
```
```suggestion
            The UUID of the newly created research data upload box
```

Cito commented on src/uos/core/orchestrator.py:
```diff
@@ -0,0 +1,392 @@
+            BoxAccessError: If the user doesn't have access to the box
+            BoxNotFoundError: If the box doesn't exist
+        """
+        # Check that the user has access to this box (if nonexistent, show unauthorized)
```
If the user is a data steward, should get the box anyway.

To this end, the method would need to take the full auth context, or have the user id optional, skipping the check if not user id is specified.

TheByronHimes replied:
Chose to do the latter

Cito commented on src/uos/core/orchestrator.py:
```diff
@@ -0,0 +1,392 @@
+        """
+        # Get existing box if user has access to it
+        user_id = UUID(auth_context.id)
+        box = await self.get_research_data_upload_box(box_id=box_id, user_id=user_id)
```
see comment below in `get_research_data_upload_box`

Cito commented on src/uos/core/models.py:
```diff
@@ -0,0 +1,258 @@
+    """An UploadGrant with the ResearchDataUploadBox title and description."""
+
+    title: str = Field(..., description="Short meaningful name for the box")
+    description: str = Field(..., description="Describes the upload box in more detail")
```
Maybe `box_title`  and `box_description` (since we already have a user name and user title, and a `box_id`)

Cito commented on src/uos/core/orchestrator.py:
```diff
@@ -0,0 +1,392 @@
+        grants_with_info: list[GrantWithBoxInfo] = []
+        for grant in grants:
+            try:
+                box = await self._box_dao.get_by_id(grant.box_id)
```
this will cause an additional request per grant, but it'll probably not help to cache their results, since one box usually has only one grant

Cito commented on src/uos/core/orchestrator.py:
```diff
@@ -0,0 +1,392 @@
+        box_id: UUID4,
+        auth_context: AuthContext,
+    ) -> Sequence[UUID4]:
+        """Get list of file IDs for an upload box.
```
```suggestion
        """Get list of file IDs for a research data upload box.
```

Cito commented on src/uos/core/orchestrator.py:
```diff
@@ -0,0 +1,392 @@
+        except ResourceNotFoundError as err:
+            raise self.BoxNotFoundError(box_id=box_id) from err
+
+        is_data_steward = "data_steward" in auth_context.roles
```
this check should be moved to a central function `is_steward` like in `auth_service.rest.auth`

Cito replied:
also, this should actually call `self.get_research_data_upload_box` to check access and get the box

Cito commented on src/uos/core/orchestrator.py:
```diff
@@ -0,0 +1,392 @@
+            )
+
+            # Get all boxes and filter to only accessible ones
+            all_boxes = [x async for x in self._box_dao.find_all(mapping={})]
```
If you fetch all boxes anyway, then you should do it at the top for both cases.

But a better strategy for a fixed user might be to call `find_one` for all ` accessible_box_ids` since these will be usually very few. so you don't need to download all the boxes (including many old, closed ones).

Cito commented on src/uos/core/tokens.py:
```diff
@@ -0,0 +1,59 @@
+ACCESS_TOKEN_LENGTH = 24
+
+
+def generate_work_order_token() -> str:
```
this function and the next one don't seem to be used

Cito reviewed (commented):
Reviewed everything except tests.

Looks good in principle, but I made several suggestions and requests for changes.

mephenor commented on src/uos/core/models.py:
```diff
@@ -175,19 +180,14 @@ class UpdateUploadBoxRequest(BaseModel):
-        description="End date of validity",
-        examples=["2023-12-31T23:59:59Z"],
-    )
+    valid_from: UTCDatetime = Field(default=..., description="Start date of validity")
+    valid_until: UTCDatetime = Field(default=..., description="End date of validity")
+    user_id: UUID4 = Field(..., description="ID of the user to grant access to")
+    iva_id: UUID4 = Field(..., description="ID of the IVA verification")
+    box_id: UUID4 = Field(..., description="ID of the upload box")
```
Probably missed that in other places, but shouldn't this be `default=...` consistently?

TheByronHimes replied:
You're right, I think I'd copy/pasted that and missed the inconsistency. I think it's only a type-check issue for us in the case of config fields, so I'll remove it from here and do a sweep for other cases.

Cito commented on src/uos/adapters/outbound/http.py:
```diff
@@ -0,0 +1,376 @@
+class FileBoxClientConfig(BaseSettings):
+    """Config parameters for interacting with the service owning FileUploadBoxes."""
+
+    file_box_service_url: HttpUrl = Field(
```
This could still be called `ucs_service_url` so that devops know what to configure.

My point was that only the FileBoxClient adapter itself needs to know that it's the UCS, not the user of the FlieBoxClient, therefore it should not be called UCSClient.

TheByronHimes replied:
I've switched it back to ucs_url. I agree that it is good to keep it like that to minimize confusion for maintaining config.

Cito commented on tests/fixtures/in_mem_dao.py:
```diff
@@ -0,0 +1,131 @@
+# limitations under the License.
+
+"""A dummy DAO"""
+
```
this was improved in UCS already and we agreed to move it to hexkit

TheByronHimes replied:
Yes, and we will. I have applied the latest UCS changes now for this mock dao class

Cito commented on tests/unit/test_orchestrator.py:
```diff
@@ -0,0 +1,628 @@
+    assert box.state == "open"
+
+
+async def test_update_research_data_upload_box_happy(rig: JointRig):
```
I think we also need an unhappy test (e.g. user tries to update title or unlock).

TheByronHimes replied:
Deal, added.

Cito commented on tests/unit/test_access_client.py:
```diff
@@ -0,0 +1,168 @@
+        await access_client.check_box_access(user_id=TEST_USER_ID, box_id=TEST_BOX_ID)
+
+    # Check network error (timeout, connection error, etc.)
+    import httpx
```
why is the import not at the top?

TheByronHimes replied:
That's just my failure to clean up generated code. Sorry you had to see that.

Cito commented on tests/unit/test_ucs_client.py:
```diff
@@ -0,0 +1,98 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Unit tests for the file box client"""
```
The file itself should also be renamed (still `test_ucs_client`).

TheByronHimes commented:
@Cito I have added the requested sorting mechanisms, pulled in the actual event schema and config options and removed the local model definitions (for the ones that are now defined in the schema library).
